### PR TITLE
ethclient/gethclient: add tracing apis to gethclient

### PR DESCRIPTION
**Pull request by bnovil**
Add methods to `ethclient/gethclient`:

- TraceCall
- TraceTransaction
- TraceChain
- TraceBlock
- TraceBlockByNumber
- TraceBlockByHash

to resolve the issue: #28182

### CODE REVIEW

File: ethclient/gethclient/gethclient_test.go [hunk 1]
```diff
@@ -518,3 +562,290 @@ func testCallContractWithBlockOverrides(t *testing.T, client *rpc.Client) {
...
+			},
+			expectErr: nil,
+			expect: map[string]interface{}{
+				"gas":         21000,
+				"failed":      false,
+				"returnValue": "",
+				"structLogs":  nil,
+			},
+		},
+	}
```
Comment (s1na) [inline ethclient/gethclient/gethclient_test.go:601]: Please also add a case for another tracer, e.g. `{ "tracer": "callTracer" }`.

File: ethclient/gethclient/gethclient.go [hunk 2]
```diff
@@ -204,6 +220,46 @@ func (ec *Client) SubscribePendingTransactions(ctx context.Context, ch chan<- co
...
+}
+
+// TraceTransaction returns the structured logs created during the execution of EVM
+func (ec *Client) TraceTransaction(ctx context.Context, hash common.Hash, config *tracers.TraceConfig) (interface{}, error) {
+	var result interface{}
+	err := ec.c.CallContext(ctx, &result, "debug_traceTransaction", hash, config)
+	return result, err
+}
+
+// TraceChain TraceChaiin subscribes to chain, receiving results from channel BlockTraceResult
```
Comment (s1na) [inline ethclient/gethclient/gethclient.go:237]:
```suggestion
// TraceChain subscribes to chain, receiving results from channel BlockTraceResult
```

File: ethclient/gethclient/gethclient_test.go (hunk 1 above)
Comment (bnovil) [reply ethclient/gethclient/gethclient_test.go:601]: ok

File: ethclient/gethclient/gethclient.go [hunk 3]
```diff
@@ -78,6 +80,20 @@ type StorageResult struct {
...
+	TxHash common.Hash `json:"txHash"`           // transaction hash
+	Result interface{} `json:"result,omitempty"` // Trace results produced by the tracer
+	Error  string      `json:"error,omitempty"`  // Trace failure produced by the tracer
+}
+
+// BlockTraceResult is the result of TraceChain
+type BlockTraceResult struct {
+	Block  hexutil.Uint64 `json:"block"`  // Block number corresponding to this trace
+	Hash   common.Hash    `json:"hash"`   // Block hash corresponding to this trace
+	Traces []interface{}  `json:"traces"` // Trace results produced by the task
```
Comment (s1na) [inline ethclient/gethclient/gethclient.go:94]:
```suggestion
	Traces []*TxTraceResult  `json:"traces"` // Trace results produced by the task
```

File: ethclient/gethclient/gethclient.go [hunk 4]
```diff
@@ -204,6 +220,46 @@ func (ec *Client) SubscribePendingTransactions(ctx context.Context, ch chan<- co
 	return ec.c.EthSubscribe(ctx, ch, "newPendingTransactions")
 }
 
+// TraceCall lets you trace a given eth_call
+func (ec *Client) TraceCall(ctx context.Context, args ethapi.TransactionArgs, blockNrOrHash rpc.BlockNumberOrHash, config *tracers.TraceCallConfig) (interface{}, error) {
```
Comment (s1na) [inline ethclient/gethclient/gethclient.go:224]: `ethapi.TransactionArgs` is in an `internal` directory which means users cannot import and use it.

```suggestion
func (ec *Client) TraceCall(ctx context.Context, args ethereum.CallMsg, blockNrOrHash rpc.BlockNumberOrHash, config *tracers.TraceCallConfig) (interface{}, error) {
```

File: ethclient/gethclient/gethclient.go [hunk 5]
```diff
@@ -204,6 +220,46 @@ func (ec *Client) SubscribePendingTransactions(ctx context.Context, ch chan<- co
...
+	return result, err
+}
+
+// TraceChain subscribes to chain, receiving results from channel BlockTraceResult
+func (ec *Client) TraceChain(ctx context.Context, ch chan<- *BlockTraceResult, start, end rpc.BlockNumber, config *tracers.TraceConfig) (*rpc.ClientSubscription, error) {
+	return ec.c.Subscribe(ctx, "debug", ch, "traceChain", start, end, config)
+}
+
+// TraceBlock returns the structured logs created during the execution of EVM
+func (ec *Client) TraceBlock(ctx context.Context, blob hexutil.Bytes, config *tracers.TraceConfig) ([]*TxTraceResult, error) {
```
Comment (s1na) [inline ethclient/gethclient/gethclient.go:243]: We shouldn't ask users to use `hexutil.Bytes`. They should work with normal Go primitives preferrably, i.e. `[]byte` here. Then we can cast it ourselves if needed.

**Comment by bnovil** [comment]
@s1na Thanks for reviewing. All issues resolved. Please review at you convenience.

File: ethclient/gethclient/gethclient.go [hunk 6]
```diff
@@ -204,6 +219,46 @@ func (ec *Client) SubscribePendingTransactions(ctx context.Context, ch chan<- co
 	return ec.c.EthSubscribe(ctx, ch, "newPendingTransactions")
 }
 
+// TraceCall lets you trace a given eth_call
+func (ec *Client) TraceCall(ctx context.Context, msg ethereum.CallMsg, blockNrOrHash rpc.BlockNumberOrHash, config *tracers.TraceCallConfig) (interface{}, error) {
```
Comment (s1na) [inline ethclient/gethclient/gethclient.go:223]: @fjl What do you think about using `rpc.BlockNumberOrHash` here? Normally we use `*big.Int` but TraceCall can accept both number or hash. Would you say it's justified or rather 2 separate methods?

File: ethclient/gethclient/gethclient_test.go [hunk 7]
```diff
@@ -518,3 +560,303 @@ func testCallContractWithBlockOverrides(t *testing.T, client *rpc.Client) {
...
+				"gas":         21000,
+				"failed":      false,
+				"returnValue": "",
+				"structLogs":  nil,
+			},
+		},
+		{
+			txHash: testTransactionHash,
+			config: &tracers.TraceConfig{
+				TracerConfig: json.RawMessage(`{"tracer":"callTracer"}`),
```
Comment (s1na) [inline ethclient/gethclient/gethclient_test.go:602]: No this is not right. `&tracers.TraceConfig{Tracer: "callTracer"}` is correct.
`TracerConfig` field is the configuration object for the given tracer.

**Review by WonderfulFull — APPROVED** [review]

**Review by WonderfulFull — APPROVED** [review]

File: ethclient/gethclient/gethclient.go (hunk 6 above)
Comment (bnovil) [reply ethclient/gethclient/gethclient.go:223]: @fjl  Hi, what's your opinion?

**Comment by bnovil** [comment]
Hi @s1na , could the PR be merged?

File: ethclient/gethclient/gethclient.go (hunk 6 above)
Comment (s1na) [reply ethclient/gethclient/gethclient.go:223]: I think it's ok. Resolving this comment.

**Comment by s1na** [comment]
@bnovil Hey sorry for the delay here. I took another look at the PR. I would like to have another set of changes. Specifically I'd like to re-define and polish the trace config objects in the ethclient code. Biggest reason is those are internally using `ethapi.StateOverride` and `ethapi.BlockOverrides` types, however ethclient has re-defined those types. This inconsistency is not good.

We can use this chance to make the config UX a bit better organized:

- It should be easy to pick one of the built-in tracers for the name (so those should be defined as constants). But also pass in name of custom tracer.
- logger.Config: this is the configuration for the structLogger. However there is overlap with `TracerConfig` which allows configuration of every other tracer. The new config object should unify that. I.e. loggerConfig should be provided as part of `TracerConfig`.
- For state and block overrides use the types defined in gethclient

Note the backend should not be changed.

**Comment by s1na** [comment]
I had another look and think we have a chance to improve the API here, specifically around config objects. Will push some changes.

**Comment by lightclient** [comment]
@bnovil please reopen if you decide to pick this back up.

**Pull request closed without merging**